### PR TITLE
1364: Jira Issues should be resolved as "Fixed"

### PR DESCRIPTION
**Pull request by zhaosongzs**
Jira Issues will be resolved as "Fixed" regardless of the current state.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-1364](https://bugs.openjdk.org/browse/SKARA-1364): Jira Issues should be resolved as "Fixed"

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara pull/1382/head:pull/1382` \
`$ git checkout pull/1382`

Update a local copy of the PR: \
`$ git checkout pull/1382` \
`$ git pull https://git.openjdk.org/skara pull/1382/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1382`

View PR using the GUI difftool: \
`$ git pr show -t 1382`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1382.diff">https://git.openjdk.org/skara/pull/1382.diff</a>

</details>

### CODE REVIEW

**Review by erikj79 — COMMENTED** [review]
Since last we talked, I have been thinking about this some more. I think we actually do need to check the `resolution` of the issue before re-opening and resolving it. If it's already "resolved" "fixed" then we shouldn't touch the state. 

The reason for this is to try to keep the bots idempotent as much as possible. If we need to rerun the notifier on commits, then if nothing needs to happen, nothing should happen.

The challenge here is that Skara doesn't currently know how to read the resolution, so we would need to implement support for that.

**Comment by erikj79** [comment]
> Since last we talked, I have been thinking about this some more. I think we actually do need to check the `resolution` of the issue before re-opening and resolving it. If it's already "resolved" "fixed" then we shouldn't touch the state.
> 
> The reason for this is to try to keep the bots idempotent as much as possible. If we need to rerun the notifier on commits, then if nothing needs to happen, nothing should happen.
> 
> The challenge here is that Skara doesn't currently know how to read the resolution, so we would need to implement support for that.

I just checked the source for `JiraIssue` and there is already a method `isFixed` which you can use. It should do exactly what we want here.

**Comment by zhaosongzs** [comment]
> > Since last we talked, I have been thinking about this some more. I think we actually do need to check the `resolution` of the issue before re-opening and resolving it. If it's already "resolved" "fixed" then we shouldn't touch the state.
> > The reason for this is to try to keep the bots idempotent as much as possible. If we need to rerun the notifier on commits, then if nothing needs to happen, nothing should happen.
> > The challenge here is that Skara doesn't currently know how to read the resolution, so we would need to implement support for that.
> 
> I just checked the source for `JiraIssue` and there is already a method `isFixed` which you can use. It should do exactly what we want here.

Got it! I will use it.

File: bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java [hunk 1]
```diff
@@ -192,9 +192,10 @@ public void onIntegratedPullRequest(PullRequest pr, Path scratchPath, Hash hash)
 
             // If prOnly is false, this is instead done when processing commits
             if (prOnly && resolve) {
-                if (issue.state() == Issue.State.OPEN) {
+                if (issue.state() != Issue.State.OPEN) {
+                    log.info("Resolving issue which status is not open, the status is " + issue.state());
+                } else {
                     log.info("Resolving issue " + issue.id());
```
Comment (erikj79) [inline bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java:198]: No need for conditional here, we can use the same log message in regardless of current state. Also, updating assignees should also be done (as long as it's currently empty).
```suggestion
                log.info("Resolving issue " + issue.id() + " from state " + issue.state());
```

File: bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java [hunk 2]
```diff
@@ -203,6 +204,11 @@ public void onIntegratedPullRequest(PullRequest pr, Path scratchPath, Hash hash)
                         }
                     }
                 }
+                if (!issue.isFixed()) {
+                    issue.setState(Issue.State.RESOLVED);
+                } else {
+                    log.info("The issue has been already resolved");
```
Comment (erikj79) [inline bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java:210]:
```suggestion
                    log.info("The issue was already 'fixed'");
```
Same below.

File: bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java [hunk 3]
```diff
@@ -311,9 +317,10 @@ public void onNewCommits(HostedRepository repository, Repository localRepository
                 if (!alreadyPostedComment) {
                     issue.addComment(commitNotification);
                 }
-                if (issue.state() == Issue.State.OPEN) {
+                if (issue.state() != Issue.State.OPEN) {
```
Comment (zhaosongzs) [inline bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java:320]: Hi @erikj79 , I am wondering whether we need condition here. Do we only update the assignee when the issue is open?

Comment (zhaosongzs) [reply bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java:320]: If it it not, I think I should update one assertion in IssueNotifierTests#testAltFixVersionsMatch

Comment (erikj79) [reply bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java:320]: oh, forgot to say, same thing here as in the other method. We don't need this conditional.

File: bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java [hunk 4]
```diff
@@ -192,17 +192,20 @@ public void onIntegratedPullRequest(PullRequest pr, Path scratchPath, Hash hash)
...
-                if (issue.state() == Issue.State.OPEN) {
-                    log.info("Resolving issue " + issue.id());
-                    issue.setState(Issue.State.RESOLVED);
-                    if (issue.assignees().isEmpty()) {
-                        var username = findIssueUsername(commit, scratchPath);
-                        if (username.isPresent()) {
-                            var assignee = issueProject.issueTracker().user(username.get());
-                            assignee.ifPresent(hostUser -> issue.setAssignees(List.of(hostUser)));
-                        }
+                log.info("Resolving issue " + issue.id() + " from state " + issue.state());
```
Comment (erikj79) [inline bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java:195]: We should keep the log message and the issue.setState call together. I think moving lines 203-207 up here would be best to keep the flow as similar to before as we can. Same applies in the other method.

Comment (zhaosongzs) [reply bots/notify/src/main/java/org/openjdk/skara/bots/notify/issue/IssueNotifier.java:195]: Sure

**Review by erikj79 — APPROVED** [review]

**Pull request closed without merging**